verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_alg_id/aws_cryptosdk_priv_hdr_parse_alg_id_harness.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/cryptosdk/header.h>
+#include <aws/cryptosdk/private/header.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+#include <proof_helpers/proof_allocators.h>
+#include <proof_helpers/utils.h>
+
+void aws_cryptosdk_priv_hdr_parse_alg_id_harness() {
+    /* Nondet Input */
+    struct aws_cryptosdk_hdr *hdr                  = ensure_nondet_hdr_has_allocated_members(MAX_TABLE_SIZE);
+    struct aws_byte_cursor *pcursor                = malloc(sizeof(*pcursor));
+    struct aws_cryptosdk_alg_properties *alg_props = malloc(sizeof(*alg_props));
+    uint8_t header_version;
+
+    /* Assumptions */
+    __CPROVER_assume(aws_cryptosdk_hdr_members_are_bounded(hdr, MAX_EDK_LIST_ITEMS, MAX_BUFFER_SIZE));
+
+    /* Precondition: The edk list has allocated list elements */
+    ensure_cryptosdk_edk_list_has_allocated_list_elements(&hdr->edk_list);
+    __CPROVER_assume(aws_cryptosdk_hdr_is_valid(hdr));
+    __CPROVER_assume(&hdr->alloc != NULL);
+
+    __CPROVER_assume(hdr->enc_ctx.p_impl != NULL);
+    ensure_hash_table_has_valid_destroy_functions(&hdr->enc_ctx);
+    size_t empty_slot_idx;
+    __CPROVER_assume(aws_hash_table_has_an_empty_slot(&hdr->enc_ctx, &empty_slot_idx));
+
+    __CPROVER_assume(pcursor != NULL);
+    __CPROVER_assume(aws_byte_cursor_is_bounded(pcursor, MAX_BUFFER_SIZE));
+    ensure_byte_cursor_has_allocated_buffer_member(pcursor);
+    __CPROVER_assume(aws_byte_cursor_is_valid(pcursor));
```

Comment:
It seems that these assumptions/preconditions about the EDK list, encryption context, and cursor are the same between most (if not all) of the `aws_cryptosdk_hdr_parse_*` proofs. Can we pull them out into a helper function?

Reply:
Yeah, there is a good chunk that is consistently shared. We usually keep assumptions about different data structures separate, so I think keeping the assumptions about the header and cursor apart would be good, but perhaps some sort of `header_setup` function would be beneficial. It would also make it more obvious when a proof does end up requiring different assumptions.

Reply:
I've made a PR for a helper function here: https://github.com/aws/aws-encryption-sdk-c/pull/664. @alex-chew @feliperodri @adpaco-aws Take a look and let me know what you think. If we're all on board with using the helper function, let's merge that PR and I'll update these proofs.